hal/interfaces/SerialCommunication.cpp

```diff
@@ -0,0 +1,40 @@
+#include "hal/interfaces/SerialCommunication.hpp"
+
+namespace hal
+{
+    BufferedSerialCommunicationOnUnbuffered::BufferedSerialCommunicationOnUnbuffered(infra::AtomicByteDeque& buffer, SerialCommunication& delegate)
+        : buffer(buffer)
+        , delegate(delegate)
+    {
+        delegate.ReceiveData([this](infra::ConstByteRange data)
+            {
+                this->buffer.Push(data);
+                scheduler.Schedule([this]()
+                    {
+                        if (HasObserver())
+                            GetObserver().DataReceived();
+                    });
+            });
+    }
+
+    BufferedSerialCommunicationOnUnbuffered::~BufferedSerialCommunicationOnUnbuffered()
+    {
+        delegate.ReceiveData([](infra::ConstByteRange) {});
```

Comment:
This does _not_ stop the delegate from actually receiving data (disabling the receive function), shouldn't the drivers implement something like 'if receivedata is called with an empty function, disable the receiver'?

Reply:
Our current SerialCommunication interfaces actually have no concept of not receiving. I'll think about if and how this should be added.